### ISSUE TITLE
just doesnt work 1.21

### ISSUE THREAD

**Issue by Saxostevesaxo** [Thomilist/dimensional-inventories#17]
The mod doesnt seem to work at all. Used this in the previous version and it was fine.
using the latest 1.21 version

I have created 3 new worlds using the multiworld mod, i can teleport to those worlds fine.
I have 3 creative worlds, which are in one pool set to creative gamemode in your mod and also set not to increment stats etc, 
and i have the default pool with default settings.

I spawn into the default pool with all my items in survival, fine.
i use a warp command to teleport to the creative dimension, I correctly lose my items but its not in creative mode.
i warp back to the default survival pool and i dont have any items... still in survival
I go through a nether portal, my items reappear and i can then go back to the overworld with my items...
warp back to the creative world, and still in survival no items, the loop continues.

the worlds also seem to take so long to load with this mod installed. looking around 30 - 60 seconds to load into the world. just displays loading terrain, without the mod installed its fine with instant tp.

<img width="197" alt="Screenshot 2024-07-27 163504" src="https://github.com/user-attachments/assets/39e7d55b-ae6a-4494-84f1-3e0b4b99ad4a">
<img width="134" alt="Screenshot 2024-07-27 163640" src="https://github.com/user-attachments/assets/557713a9-209c-48fe-8607-8a707efc3f01">

**Comment by Thomilist** [comment]
I can't seem to find a compatible set of mod versions to try and reproduce this. Can you list the mods you are using here? Maybe with links? Because [Multiworld](https://modrinth.com/mod/multiworld/version/1.8.1) seems to only be out for 1.21-pre2 at this point, not the full 1.21 release, as far as I can tell.

If you can attach a log file as well, that would be great.

**Comment by Saxostevesaxo** [comment]
maybe its multiworld that the issue then?

i'm using a large set of mods to be honest, i have this version of multiworld if you wanted to try with that?

[Multiworld-Fabric-1.21.zip](https://github.com/user-attachments/files/16400955/Multiworld-Fabric-1.21.zip)

I also use multi world borders, 
<img width="149" alt="Screenshot 2024-07-27 181049" src="https://github.com/user-attachments/assets/3e5b79a1-4492-4a69-9431-bfae4495de30">
<img width="127" alt="Screenshot 2024-07-27 181116" src="https://github.com/user-attachments/assets/eded018c-5334-48b3-aa09-e73b23f5437e">

**Comment by Saxostevesaxo** [comment]
[2024-07-27-7.log](https://github.com/user-attachments/files/16401087/2024-07-27-7.log)
this was latest log, no errors from your mod that i can see

**Comment by Thomilist** [comment]
Thank you.

I still haven't been able to reproduce the issue, unfortunately. I have tried with just `Dimensional Inventories` + `Multiworld`, which worked as expected. I have also tried with almost all of the mods you listed ("almost" because some of them gave unrelated errors that either prevented me from launching the server or from connecting to it. Missing were `owo lib`, `Deathlog`, `LuckPerms-Fabric-PlaceholderAPI-Hook`, `Geyser` and `Floodgate`).

That being said, I have a hunch that it might be related to either `C2ME` or `Very Many Players`. The slow load times you mention sound very similar to my own experience some time last year. I believe it went away after I removed `C2ME`, so you could give that a try. Might narrow it down, if nothing else.

Quite a few of the mods you use have also been updated since you downloaded them. I used the latest versions available (except with `C2ME`, for which I tried both `0.2.0+alpha.11.107` (latest) and `0.2.0+alpha.11.104` (the one you used), so perhaps something got fixed in the meantime?

**Comment by Saxostevesaxo** [comment]
ok, thanks for taking a look, very strange, I will try updating the mods I have and let you know as soon a s i can, might be another couple of days before I can take a deep look at it.

**Comment by Gagabolol** [comment]
I have the exact same problem and i don't have C2ME or Very many players

**Comment by Saxostevesaxo** [comment]
I have updated all of my mods and it still has the same issue, I have a test server running if you wanted to join and see how it behaves...?

135.148.168.188:25569

you can use /warp tp
to TP between worlds.
If you let me know your MC username I can op you, and if you want a copy of my backup file let me know and I can get it to you somehow.

I have also added permissions to set game mode on the default permissions.

spotted another issue actually...

when you tp to the creative world and give yourself some items and then go back to the normal survival, you will have the creative inventory...
but if you tp to the creative world and don't give yourself any items and go back to survival you will get your survival items back but only after going through the nether portal... strange happenings :)

thanks for looking into it tho!

**Comment by Saxostevesaxo** [comment]
I have added permissions to op yourself if you like

**Comment by Saxostevesaxo** [comment]
also just had a crash caused by the mod I think 

[crash-2024-07-28_23.18.02-server.txt](https://github.com/user-attachments/files/16405535/crash-2024-07-28_23.18.02-server.txt)

**Comment by Thomilist** [comment]
> also just had a crash caused by the mod I think
> 
> [crash-2024-07-28_23.18.02-server.txt](https://github.com/user-attachments/files/16405535/crash-2024-07-28_23.18.02-server.txt)

I think I've managed to reproduce it now. The crash log was certainly a good hint.

Enchanted items, and possibly other non-plain items, cause an exception when trying to serialise items as NBT data. Item code, and specifically item <-> NBT conversion code, was changed significantly in Minecraft 1.20.5 - and indeed, the issue is not present in 1.20.4.

So it *is* caused by Dimensional Inventories, not other mods. Thank you for helping me narrow it down. It should be fixable at this point.

**Comment by ScorNinja** [comment]
Yes. I just tested and recreated this issue with just an enchanted sword with NBT data. All the items reappear after opening a chest.

**Comment by Thomilist** [comment]
How about this one: [dimensional-inventories-2.0.2-dev.1+1.21.zip](https://github.com/user-attachments/files/16406048/dimensional-inventories-2.0.2-dev.1%2B1.21.zip)

**Comment by ScorNinja** [comment]
> How about this one: [dimensional-inventories-2.0.2-dev.1+1.21.zip](https://github.com/user-attachments/files/16406048/dimensional-inventories-2.0.2-dev.1%2B1.21.zip)

This appears to work for me including custom swords and custom items and bundles and everything.

**Comment by Saxostevesaxo** [comment]
That seems to have fixed all the issues :)

**Comment by Thomilist** [comment]
Alright, nice. Thank you for all the helpful efforts. I'll wrap it into a proper release, and then probably look into some automated testing to cover edge cases more consistently.

**Comment by Thomilist** [comment]
Patch released for [1.20.6](https://github.com/Thomilist/dimensional-inventories/releases/tag/2.0.2%2B1.20.6) and [1.21](https://github.com/Thomilist/dimensional-inventories/releases/tag/2.0.2%2B1.21).

The bug affected code introduced for 1.20.5+ compatibility, so it "only" applied to versions `2.0.0+1.20.6`, `2.0.1+1.20.6`, `2.0.0+1.21` and `2.0.1+1.21`. Aside from the names, `2.0.2-dev.1+1.21` and `2.0.2+1.21` are identical.
